### PR TITLE
[FLINK-25640][docs] Enhance the document for blocking shuffle

### PR DESCRIPTION
**Pull request by wsry**
## What is the purpose of the change

Enhance the document for blocking shuffle and add more details about performance tuning and trouble shooting.

## Brief change log

  - Enhance the document for blocking shuffle and add more details about performance tuning and trouble shooting.

## Verifying this change

Document change without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by gaoyunhaii** [comment]
@flinkbot run azure

File: docs/content.zh/docs/ops/batch/blocking_shuffle.md [hunk 1]
```diff
@@ -56,34 +56,59 @@ Flink [DataStream API]({{< ref "docs/dev/datastream/execution_mode" >}}) 和 [Ta
...
+1. 如果你使用机械硬盘作为存储设备，请总是使用 `Sort Shuffle`，因为这可以极大的提升稳定性和性能。从 1.15 开始，`Sort Shuffle` 已经成为默认实现，对于 1.14 以及更低版本，你需要通过将 [taskmanager.network.sort-shuffle.min-parallelism]({{< ref "docs/deployment/config" >}}#taskmanager-network-sort-shuffle-min-parallelism) 配置为 1 以手动开启 `Sort Shuffle`。
+2. 对于 `Sort Shuffle` 和 `Hash Shuffle` 两种实现，你都可以考虑开启 [数据压缩]({{< ref "docs/deployment/config">}}#taskmanager-network-blocking-shuffle-compression-enabled) 除非数据本身无法压缩。从 1.15 开启，数据压缩是默认开启的，对于 1.14 以及更低版本你需要手动开启。
+3. 当使用 `Sort Shuffle` 时， 减少[独占网络缓冲区]({{< ref "docs/deployment/config" >}}#taskmanager-network-memory-buffers-per-channel) 并增加[流动网络缓冲区]({{< ref "docs/deployment/config" >}}#taskmanager-network-memory-floating-buffers-per-gate) 有利于性能提升。对于 1.14 以及更高版本，建议将 [taskmanager.network.memory.buffers-per-channel]({{< ref "docs/deployment/config" >}}#taskmanager-network-memory-buffers-per-channel) 设为 0 并且将 [taskmanager.network.memory.floating-buffers-per-gate]({{< ref "docs/deployment/config" >}}#taskmanager-network-memory-floating-buffers-per-gate) 设为一个较大的值 (比如, 4096)。 这有两个主要的好处：1) 首先这解耦了并发与网络内存使用量，对于大规模作业，这降低了遇到 "Insufficient number of network buffers" 错误的可能性；2) 网络缓冲区可以根据需求在不同的数据通道间共享流动，这可以提高了网络缓冲区的利用率，进而可以提高性能。
```
Comment (gaoyunhaii) [inline docs/content.zh/docs/ops/batch/blocking_shuffle.md:94]: `减少[独占网络缓冲区]` 前面多一个空格。
[独占网络缓冲区]和[流动网络缓冲区] 前后的空格最好保持一致，现在是后面有前面没有，下面的链接也类似

File: docs/content.zh/docs/ops/batch/blocking_shuffle.md [hunk 2]
```diff
@@ -56,34 +56,59 @@ Flink [DataStream API]({{< ref "docs/dev/datastream/execution_mode" >}}) 和 [Ta
...
+| Connection reset by peer | 这通常意味着网络不太稳定或者压力较大。其他一些原因，比如上面提到的 SSL 握手超时等也可能会导致这一问题。如果你使用的是 `Hash Shuffle`，请切换到 `Sort Shuffle`。如果你已经在使用 `Sort Shuffle`，增大[网络连接 backlog]({{< ref "docs/deployment/config" >}}#taskmanager-network-netty-server-backlog) 可能会有所帮助。|
+| Network connection timeout | 这通常意味着网络不太稳定或者压力较大。增大[网络连接超时时间]({{< ref "docs/deployment/config" >}}#taskmanager-network-netty-client-connectTimeoutSec) 或者开启[网络连接重试]({{< ref "docs/deployment/config" >}}#taskmanager-network-retries) 可能会有所帮助。|
+| Socket read/write timeout | 这通常意味着网络传输速度较慢或者压力较大。增大[网络收发缓冲区]({{< ref "docs/deployment/config" >}}#taskmanager-network-netty-sendReceiveBufferSize) 大小可能会有所帮助。如果作业运行在 Kubernetes 环境，使用 [host network]({{< ref "docs/deployment/config" >}}#kubernetes-hostnetwork-enabled) 可能会有所帮助。|
+| Read buffer request timeout | 这个问题只会出现在 `Sort Shuffle`，它意味着对数据读取缓冲区的激烈竞争。要解决这一问题，你可以增大 [taskmanager.memory.framework.off-heap.batch-shuffle.size]({{< ref "docs/deployment/config" >}}#taskmanager-memory-framework-off-heap-batch-shuffle-size) 和 [taskmanager.memory.framework.off-heap.size]({{< ref "docs/deployment/config" >}}#taskmanager-memory-framework-off-heap-size). |
```
Comment (gaoyunhaii) [inline docs/content.zh/docs/ops/batch/blocking_shuffle.md:110]: 可能可以改用中文句号

File: docs/content.zh/docs/ops/batch/blocking_shuffle.md [hunk 3]
```diff
@@ -56,34 +56,59 @@ Flink [DataStream API]({{< ref "docs/dev/datastream/execution_mode" >}}) 和 [Ta
...
+| No space left on device | 这通常意味着磁盘存储空间或者 inodes 被耗尽。你可以考虑扩展磁盘存储空间或者做一些数据清理。|
+| Out of memory error | 如果你使用的是 `Hash Shuffle`，请切换到 `Sort Shuffle`。如果你已经在使用 `Sort Shuffle` 并且遵循了上面章节的建议，你可以考虑增大相应的内存大小。对于堆上内存，你可以增大 [taskmanager.memory.task.heap.size]({{< ref "docs/deployment/config" >}}#ttaskmanager-memory-task-heap-size)，对于直接内存，你可以增大 [taskmanager.memory.task.off-heap.size]({{< ref "docs/deployment/config" >}}#taskmanager-memory-task-off-heap-size). |
```
Comment (gaoyunhaii) [inline docs/content.zh/docs/ops/batch/blocking_shuffle.md:112]: 可能可以改用中文句号

File: docs/content/docs/ops/batch/blocking_shuffle.md [hunk 4]
```diff
@@ -59,31 +59,55 @@ If [SSL]({{< ref "docs/deployment/security/security-ssl" >}}) is enabled, the `f
...
 ## Choices of Blocking Shuffle
 
 As a summary,
 
 - For small scale jobs running on SSD, both implementation should work.
-- For large scale jobs or for jobs running on HDD, `sort shuffle` should be more suitable.
-- In both case, you may consider [enabling compression]({{< ref "docs/deployment/config">}}#taskmanager-network-blocking-shuffle-compression-enabled) to improve the performance unless the data is hard to compress.
+- For large scale jobs or for jobs running on HDD, `Sort Shuffle` should be more suitable.
+
+To switch between `Sort Shuffle` and `Hash Shuffle`, you need to adjust this config option: [taskmanager.network.sort-shuffle.min-parallelism]({{< ref "docs/deployment/config" >}}#taskmanager-network-sort-shuffle-min-parallelism). It controls which shuffle implementation to use based on the parallelism of downstream tasks, if the parallelism is lower than the configured value, `Hash Shuffle` will be used, otherwise `Sort Shuffle` will be used. For versions lower than 1.15, its default value is `Integer.MAX_VALUE`, so `Hash Shuffle` will be always used by default. Since 1.15, its default value is 1, so `Sort Shuffle` will be always used by default.
```
Comment (gaoyunhaii) [inline docs/content/docs/ops/batch/blocking_shuffle.md:86]: will be always used by default -> will be used by default ?

File: docs/content/docs/ops/batch/blocking_shuffle.md [hunk 5]
```diff
@@ -59,31 +59,55 @@ If [SSL]({{< ref "docs/deployment/security/security-ssl" >}}) is enabled, the `f
...
+
+## Performance Tuning
+
+The following guidelins may help you to achieve better performance especially for large scale batch jobs:
+
+1. Always use `Sort Shuffle` on spinning disk because `Sort Shuffle` can largely improve stability and IO performance. Since 1.15, `Sort Shuffle` is already the default blocking shuffle implementation, for 1.14 and lower version, you need to enable it manually by setting [taskmanager.network.sort-shuffle.min-parallelism]({{< ref "docs/deployment/config" >}}#taskmanager-network-sort-shuffle-min-parallelism) to 1.
```
Comment (gaoyunhaii) [inline docs/content/docs/ops/batch/blocking_shuffle.md:92]: Might unify `spinning disk` -> `HDD` or reservely

File: docs/content/docs/ops/batch/blocking_shuffle.md [hunk 6]
```diff
@@ -59,31 +59,55 @@ If [SSL]({{< ref "docs/deployment/security/security-ssl" >}}) is enabled, the `f
...
+3. When `Sort Shuffle` is used, decrease the number of [exclusive buffers per channel]({{< ref "docs/deployment/config" >}}#taskmanager-network-memory-buffers-per-channel) and increase the number of [floating buffers per gate]({{< ref "docs/deployment/config" >}}#taskmanager-network-memory-floating-buffers-per-gate) can help. For 1.14 and higher version, it is suggested to set [taskmanager.network.memory.buffers-per-channel]({{< ref "docs/deployment/config" >}}#taskmanager-network-memory-buffers-per-channel) to 0 and set [taskmanager.network.memory.floating-buffers-per-gate]({{< ref "docs/deployment/config" >}}#taskmanager-network-memory-floating-buffers-per-gate) to a larger value (for example, 4096). This setting has two main advantages: 1) It decouples the network memory consumption from parallelism so for large scale jobs, the possibility of "Insufficient number of network buffers" error can be decreased; 2) Networker buffers are distributed among different channels according to needs which can improve the network buffer utilization, thus can improve performance.
```
Comment (gaoyunhaii) [inline docs/content/docs/ops/batch/blocking_shuffle.md:94]: decrease the number of-> decreasing the number of
increase the number of -> increasing the number of

`...according to needs which can improve the network buffer utilization, thus can improve performance.` -> 
`...according to needs, which can improve the network buffer utilization and further improve the performance.`

File: docs/content/docs/ops/batch/blocking_shuffle.md [hunk 7]
```diff
@@ -59,31 +59,55 @@ If [SSL]({{< ref "docs/deployment/security/security-ssl" >}}) is enabled, the `f
...
+
+## Trouble Shooting
+
+Here are some exceptions you may encounter (rarely) and the corresponding solutions that may help:
+
+| Exceptions | Potential Solutions |
+| :--------- | :------------------ |
+| Insufficient number of network buffers | This means the amount of network memory is not enough to run the target job and you need to increase the total network memory size. Note that since 1.15, `Sort Shuffle` has become the default blocking shuffle implementation and for some cases, it may need more network memory than before, which means there is a small possibility that your batch jobs may suffer this issue after upgrading to 1.15. If this is the case, you just need to increase the total network memory size. |
```
Comment (gaoyunhaii) [inline docs/content/docs/ops/batch/blocking_shuffle.md:105]: suffer this issue -> suffer from this issue ?

File: docs/content/docs/ops/batch/blocking_shuffle.md [hunk 8]
```diff
@@ -59,31 +59,55 @@ If [SSL]({{< ref "docs/deployment/security/security-ssl" >}}) is enabled, the `f
...
+| Too many open files | This means that there is no enough file descriptors left. If you are using `Hash Shuffle`, please switch to `Sort Shuffle`. If you are already using `Sort Shuffle`, please consider to increase the system limit for file descriptor and check if the user code consumes too many file descriptors. |
```
Comment (gaoyunhaii) [inline docs/content/docs/ops/batch/blocking_shuffle.md:106]: no enough -> not enough ?
consider to increase -> consider increasing?

File: docs/content/docs/ops/batch/blocking_shuffle.md [hunk 9]
```diff
@@ -59,31 +59,55 @@ If [SSL]({{< ref "docs/deployment/security/security-ssl" >}}) is enabled, the `f
...
+| Connection reset by peer | This usually means that the network if unstable or or under heavy burden. Other issues like SSL handshake timeout mentioned above may also cause this problem. If you are using `Hash Shuffle`, please switch to `Sort Shuffle`. If you are already using `Sort Shuffle`, increasing the [network backlog]({{< ref "docs/deployment/config" >}}#taskmanager-network-netty-server-backlog) may help. |
```
Comment (gaoyunhaii) [inline docs/content/docs/ops/batch/blocking_shuffle.md:107]: if unstable -> is unstable

File: docs/content/docs/ops/batch/blocking_shuffle.md [hunk 10]
```diff
@@ -59,31 +59,55 @@ If [SSL]({{< ref "docs/deployment/security/security-ssl" >}}) is enabled, the `f
...
+| Network connection timeout | This usually means that the network if unstable or or under heavy burden and increasing the [network connection timeout]({{< ref "docs/deployment/config" >}}#taskmanager-network-netty-client-connectTimeoutSec) or enable [connection retry]({{< ref "docs/deployment/config" >}}#taskmanager-network-retries) may help. |
```
Comment (gaoyunhaii) [inline docs/content/docs/ops/batch/blocking_shuffle.md:108]: if unstable or or under heavy burden -> is unstable or under heavy burden

File: docs/content/docs/ops/batch/blocking_shuffle.md [hunk 11]
```diff
@@ -59,31 +59,55 @@ If [SSL]({{< ref "docs/deployment/security/security-ssl" >}}) is enabled, the `f
...
+| Socket read/write timeout | This may indicate that the network is slow or under heavy burden and increase the [network send/receive buffer size]({{< ref "docs/deployment/config" >}}#taskmanager-network-netty-sendReceiveBufferSize) may help. If the job is running in Kubernetes environment, using [host network]({{< ref "docs/deployment/config" >}}#kubernetes-hostnetwork-enabled) may also help. |
```
Comment (gaoyunhaii) [inline docs/content/docs/ops/batch/blocking_shuffle.md:109]: increase the -> increasing the

**Review by gaoyunhaii — COMMENTED** [review]
Very thanks @wsry for the PR! I have left some comments.

Also, currently it seems Sort shuffle have several options to control the memory usage and they are also related. Might you add a concise description in the `Sort shuffle` section about the relationship between these options?

**Comment by wsry** [comment]
@gaoyunhaii Thanks for the review and feedback. I have updated the PR accordingly.

**Review by gaoyunhaii — APPROVED** [review]
Very thanks @wsry for the update! LGTM

**Pull request closed without merging**